### PR TITLE
Fixes a couple issues with the AzureBlobStorage implementation

### LINKED ISSUE #433: Concurrency issue in AzureBlobStorage::Read
The implementation of `AzureBlobStorage::Read` has a concurrency issue in that it spins off multiple async which can each complete concurrently and, once completed, the logic modifies a single shared `Dictionary<K, V>` instance.

Line 138 is the problem spot here:

https://github.com/Microsoft/botbuilder-dotnet/blob/88c9242112f1ca6b1e8528fc05f34e141d784c08/libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs#L113-L151

Instead we should spin off the reads asynchronously and then each one should produce its own `KeyValuePair<string, object>` which the main code path can `Task::WhenAll` on and return as a unified result.

### LINKED ISSUE #434: AzureBlobStorage use of Lazy<> to ensure CloudBlobContainer is still problematic
`AzureBlobStorage` is now using a `Lazy<ValueTask<CloudBlobContainer>>`, but this doesn't ultimately achieve the goals. First, it effectively eliminates any gain of using `ValueTask<T>` because it will 100% of the time be wrapping a `Task<CloudBlobContainer>` and never a raw `CloudBlobContainer`. So, while this pattern _does work_, it might as well just be typed as `Lazy<Task<CloudBlobContainer>>`.

That said, I would still suggest a factory method approach around a field that we can simply `await` like any other method and then that method actually can use a `ValueTask<CloudBlobContainer>` to optimize all calls except for the first that actually needs to resolve the container to allocate no `Task<CloudBlobContainer>` at all and, with the use of a local function, we can avoid the overhead of any async state machine in those cases as well.

### PR DESCRIPTION
**Pull request by drub0y**
Fixes #433 
Fixes #434

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs [hunk 1]
```diff
@@ -174,12 +167,40 @@ private static string GetBlobName(string key)
...
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        private ValueTask<CloudBlobContainer> GetBlobContainer()
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs]: >ValueTask [](start = 16, length = 9)

This is cool, I hadn't heard of this. I will have to start using this.

File: libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs [hunk 2]
```diff
@@ -174,12 +167,40 @@ private static string GetBlobName(string key)
...
+        {
+            if(_container != null)
+            {
+                return new ValueTask<CloudBlobContainer>(_container);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs]: >new ValueTask<CloudBlobContainer>( [](start = 23, length = 34)

wait a minute, the docs I read said you should be able to return the _container itself here.  Did I read the wrong docs?

**Comment by cleemullins** [comment]
@drub0y - this looks great, but for some reason the build isn't passing. 

Let me look at that, and then I'll merge.

**Comment by drub0y** [comment]
@cleemullins Yeah, it was blowing up on the Azure Blob Transcript tests, but I think I saw something from @tomlm go in that should have fixed that. Let me rebase so we can get a clean build here.

File: libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs (hunk 2 above)
Comment (drub0y) [reply libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs]: Yeah, you'll notice I didn't use the `async` keyword on the method, so I'm in charge of having to construct the `ValueTask<T>` manually myself in that case. Here's the difference in IL:

Without `async` keyword:

```MSIL
.method private hidebysig instance valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> GetBlobContainer () cil managed 
    {
        .locals init (
            [0] bool V_0,
            [1] valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> V_1
        )

        IL_0000: nop
        IL_0001: ldarg.0
        IL_0002: ldfld class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer Microsoft.Bot.Builder.Azure.AzureBlobStorage::_container
        IL_0007: ldnull
        IL_0008: cgt.un
        IL_000a: stloc.0
        IL_000b: ldloc.0
        IL_000c: brfalse.s IL_001d

        IL_000e: nop
        IL_000f: ldarg.0
        IL_0010: ldfld class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer Microsoft.Bot.Builder.Azure.AzureBlobStorage::_container
        IL_0015: newobj instance void valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::.ctor(!0)
        IL_001a: stloc.1
        IL_001b: br.s IL_002b

        IL_001d: ldarg.0
        IL_001e: call instance class [netstandard]System.Threading.Tasks.Task`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage::'<GetBlobContainer>g__EnsureBlobContainerExists|10_0'()
        IL_0023: newobj instance void valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::.ctor(class [netstandard]System.Threading.Tasks.Task`1<!0>)
        IL_0028: stloc.1
        IL_0029: br.s IL_002b

        IL_002b: ldloc.1
        IL_002c: ret
    }
```

With `async` keyword:

```MSIL
.method private hidebysig instance valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> GetBlobContainer () cil managed 
    {
        .custom instance void [netstandard]System.Runtime.CompilerServices.AsyncStateMachineAttribute::.ctor(class [netstandard]System.Type) = (
            01 00 44 4d 69 63 72 6f 73 6f 66 74 2e 42 6f 74
            2e 42 75 69 6c 64 65 72 2e 41 7a 75 72 65 2e 41
            7a 75 72 65 42 6c 6f 62 53 74 6f 72 61 67 65 2b
            3c 47 65 74 42 6c 6f 62 43 6f 6e 74 61 69 6e 65
            72 3e 64 5f 5f 31 30 00 00
        )
        .custom instance void [netstandard]System.Diagnostics.DebuggerStepThroughAttribute::.ctor() = (
            01 00 00 00
        )
        .locals init (
            [0] class Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10' V_0,
            [1] valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> V_1
        )

        IL_0000: newobj instance void Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::.ctor()
        IL_0005: stloc.0
        IL_0006: ldloc.0
        IL_0007: ldarg.0
        IL_0008: stfld class Microsoft.Bot.Builder.Azure.AzureBlobStorage Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>4__this'
        IL_000d: ldloc.0
        IL_000e: call valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::Create()
        IL_0013: stfld valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>t__builder'
        IL_0018: ldloc.0
        IL_0019: ldc.i4.m1
        IL_001a: stfld int32 Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>1__state'
        IL_001f: ldloc.0
        IL_0020: ldfld valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>t__builder'
        IL_0025: stloc.1
        IL_0026: ldloca.s V_1
        IL_0028: ldloca.s V_0
        IL_002a: call instance void valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::Start<class Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'>(!!0&)
        IL_002f: ldloc.0
        IL_0030: ldflda valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> Microsoft.Bot.Builder.Azure.AzureBlobStorage/'<GetBlobContainer>d__10'::'<>t__builder'
        IL_0035: call instance valuetype [System.Threading.Tasks.Extensions]System.Threading.Tasks.ValueTask`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer> valuetype [System.Threading.Tasks.Extensions]System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1<class [Microsoft.WindowsAzure.Storage]Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer>::get_Task()
        IL_003a: ret
    }
```

So while the `async` one might read a little better, it's always costing you an allocation of that async state machine which I think is worthwhile optimizing away in a case like this with a little more explicit construction of the `ValueTask<T>`.

File: libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs (hunk 1 above)
Comment (drub0y) [reply libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs]: Yeah, used in the right situations (and correctly) there's a potential to save a lot of allocations by using `ValueTask<T>`. Probably one of the best additions to .NET for framework writers in a while. 🎉

**Comment by drub0y** [comment]
@cleemullins The rebase picked up the test fixes from `master`, looks like this is good to go. Hollar if you need anything else, otherwise I think we're good. :shipit:

File: libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs (hunk 2 above)
Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs]: @tomlm, take that! :) 

heh.

**Comment by cleemullins** [comment]
@drub0y Fixed.

**Comment by cleemullins** [comment]
@drub0y Merged in! Thanks!!

**Pull request merged**